Add CI Pipeline for Automated Testing

PaintedW0lf opened the pull request:
<!-- 
Thank you for contributing! Please fill out this template to help us review your PR.
-->

## 📝 Description

This PR adds automated testing infrastructure using GitHub Actions. Every time a PR is created or code is pushed to main/master, the CI pipeline will automatically run all tests and generate coverage reports. This ensures code quality and catches bugs before they reach production.
Additionally, I added the plan for further CI/CD workflow in the readme. Currently, 2-3 parts are a TODO; I will keep on adding more parts in separate PRs once we have this pipeline working.

**NOTE** : The current failure is -(and i opened the PR for that https://github.com/COSC-499-W2025/capstone-project-team-3/issues/429 - Added bug)

``ImportError while importing test module '/home/runner/work/capstone-project-team-3/capstone-project-team-3/tests/utils/test_analysis_merger.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/opt/hostedtoolcache/Python/3.11.14/x64/lib/python3.11/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
tests/utils/test_analysis_merger.py:1: in <module>
    from app.utils.analysis_merger_utils import merge_analysis_results,store_results_in_db, retrieve_results_from_db
E   ImportError: cannot import name 'retrieve_results_from_db' from 'app.utils.analysis_merger_utils' (/home/runner/work/capstone-project-team-3/capstone-project-team-3/app/utils/analysis_merger_utils.py)``

The upload permission errors have been fixed, and we can raise a bug for fixing the pre-existing import error in /test_analysis_merger.py. We can make sure the app.utils.analysis_merger_utils module exists at the correct location and is accessible from the tests directory. Not fixing the error here to stay in the scope of the PR and issue.



**Closes:** #291 , #292 

---

## 🔧 Type of Change

- [ ] 🐛 Bug fix (non-breaking change that fixes an issue)
- [x] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] 📚 Documentation added/updated
- [x] ✅ Test added/updated
- [ ] ♻️ Refactoring
- [x ] ⚡ Performance improvement

---

## 🧪 Testing

**Local Testing:**
- [x] Validated YAML syntax is correct
- [x] Tested pytest runs successfully: `pytest`
- [x] Tested coverage generation: `pytest --cov=app --cov-report=term-missing`
- [x] Verified all dependencies (pytest, pytest-cov) exist in requirements.txt

**CI Testing:**
- [ ] Workflow will run automatically when this PR is created
- [ ] Check "Actions" tab to verify workflow executes successfully
- [ ] Confirm all tests pass in CI environment

**How to verify:**
1. Once PR is created, go to the "Actions" tab
2. Look for "CI Pipeline" workflow running
3. Verify it completes with green checkmark ✅
4. Review test output and coverage report in workflow logs

---

## ✓ Checklist

- [x] 🤖 GenAI was used in generating the code and I have performed a self-review of my own code
- [x] 💬 I have commented my code where needed
- [x] 📖 I have made corresponding changes to the documentation
- [x] ⚠️ My changes generate no new warnings
- [x] ✅ I have added tests that prove my fix is effective or that my feature works and tests are passing locally
- [x] 🔗 Any dependent changes have been merged and published in downstream modules
- [ ] 📱 Any UI changes have been checked to work on desktop, tablet, and/or mobile (N/A - infrastructure only)

---

## 📸 Screenshots

> If applicable, add screenshots to help explain your changes
<img width="661" height="645" alt="Screenshot 2026-01-10 at 4 50 30 PM" src="https://github.com/user-attachments/assets/9521fe55-829e-419e-b170-bd2aaa8ebbfe" /><details>
<summary>Click to expand screenshots</summary>

6s-1 commented:
Thanks @PaintedW0lf, I haven't directly worked on the files you mentioned, but I know through group discussions they are still under development, so ideally they could have been added to the skipped tests for now. (Most CI setups, including GitHub Actions, allow certain tests to be temporarily skipped while features/files are still in progress.) Since CI/CD updates or planning didn’t come up in our previous meetings (including this Tuesday’s), it would really help if we could bring these kinds of project-wide changes to our group discussions beforehand, so you can get input on files that are still under development before they end up causing issues in the pipeline, and so that we’re all aligned moving forward.

PaintedW0lf commented:
Thanks for the comments Dabira, as mentioned in group chat -  i only added the comment here to address the changes so that reviewers dont think that they would request chnages based on the  failed check mark 

Other than that, we dont have to work these issues right away 


@6s-1  - as u know my next pr (and current work of non code analysis)  is blocked because i am still waiting on  revirews 
So in the meanwhile i started working to next steps that are equally important and informed in our group chat- Please check group chat for more information. 

Can i please know how implementing this workflow is impacting the current work ? Because if so we can always merge the pr later.  Also, since our milestone is due soon i am trying to speed up the process of work and automate some of the things as we all discussed in our team the concenrs.

PaintedW0lf commented:
Thanks for the comments Dabira, as mentioned in group chat, I only added the comment here to address the changes so that reviewers dont think that they would request changes based on the  failed checkmark. 

Other than that, we dont have to work these issues right away.. just to keep in backlog as needed

@6s-1—as you know, my next pr (and current work of non-code analysis)  is blocked because i am still waiting on reviews. 
So in the meanwhile I started working on the next steps that are equally important and informed in our group chat—please check the group chat for more information. 

Can I please know how implementing this workflow is impacting the current work? Because if so, we can always merge the PR later.  Also, since our milestone is due soon, I am trying to speed up the process of work and automate some of the things as we all discussed in our team the concenrs.

PaintedW0lf commented:
Good catch @6s-1  , thanks i fixed it and retested it to pass the tests for the change.
Removed [tmp_path](vscode-file://vscode-app/private/var/folders/7l/50q2f7k16_df1k4z6_l7ww680000gn/T/AppTranslocation/F0AA4016-2C71-4CA2-8510-5903C696B6EF/d/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) parameter - no longer needed since we get the path from the env var
Use [os.getenv("UPLOAD_DIR")](vscode-file://vscode-app/private/var/folders/7l/50q2f7k16_df1k4z6_l7ww680000gn/T/AppTranslocation/F0AA4016-2C71-4CA2-8510-5903C696B6EF/d/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) - reads the environment variable that was set in the fixture
<img width="692" height="117" alt="Screenshot 2026-01-11 at 5 36 45 PM" src="https://github.com/user-attachments/assets/c5ed41ec-4a2b-46a5-b545-b7dbc98db49c" />

6s-1 reviewed (approved):
Thank you for making these changes Vanshika, looks good to me!
